### PR TITLE
feat: add locks to make class manipulation thread safe v2

### PR DESCRIPTION
**Pull request by JuvenalDuarte**
#### Please provide details about this Pull Request (why the change is being made, what this commit will do):
Changes made to address issue 572 (https://github.com/totvslabs/pyCarol/issues/572).

### CODE REVIEW

File: pycarol/carolina.py [hunk 1]
```diff
@@ -25,49 +28,62 @@ def __init__(self, carol):
...
-            token['app_name'] = self.carol.app_name
-            Carolina.token = token
+        # locks all instance variables to prevent inconsistencies
+        with Carolina.backend_lock:
+            if self.client:
+                # Check if the token is not expired for at least another minute.. we do a little margin to avoid time difference issues
+                if self.expires_at is None or datetime.utcnow() + timedelta(minutes=1) < self.expires_at:
+                    return False
+                else:
+                    expired = True
```
Comment (rafarui) [inline pycarol/carolina.py:47]: this part does not need to be inside the thread lock.

File: pycarol/carolina.py [hunk 2]
```diff
@@ -25,49 +28,62 @@ def __init__(self, carol):
...
+                    or Carolina.token.get('app_name', '') != self.carol.app_name \
+                    or (datetime.utcnow() + timedelta(minutes=1)
+                        >= datetime.fromtimestamp(Carolina.token.get('expirationTimestamp', 1)/1000.0)):
+                token = self.carol.call_api('v1/storage/storage/token', params={'carolAppName': self.carol.app_name})
+                token['tenant_name'] = self.carol.tenant['mdmName']
+                token['app_name'] = self.carol.app_name
+                Carolina.token = token
+
+            token = Carolina.token
+            self.engine = token['engine']
```
Comment (rafarui) [inline pycarol/carolina.py:60]: from this line you can leave the lock. the rest does not need to be lock

File: pycarol/storage.py [hunk 3]
```diff
@@ -22,27 +22,33 @@ def __init__(self, carol):
...
+            True if storage object has been reinitialized, false if it is still valid.
+
+        """
+        if self.carolina.init_if_needed():
+
+            if self.carolina.engine == 'GCP-CS':
+                from .utils.storage_gcpcs import StorageGCPCS
+
+                with self.backend_lock:
+                    self.backend = StorageGCPCS(self.carol, self.carolina)
```
Comment (rafarui) [inline pycarol/storage.py:43]: you dont need this. when this storages uses the Caroline() it will be lock by the carolina itself, no need to lock here.

**Review by rafarui — CHANGES_REQUESTED** [review]

File: pycarol/carolina.py (hunk 1 above)
Comment (JuvenalDuarte) [reply pycarol/carolina.py:47]: This first `if` is the key. If two threads evaluates it concurrently in a fraction of miliseconds they could both fall into `expired = True`. In that case they would start reseting the token, engine and all the instance vars, which might cause inconsistencies. 
By the code I guess both threads would set the exact same values, but I prefer leaving the whole method as a critical zone. In that way the first thread to get the lock would reset the vars, the second would wait, evatuate `expired = False` and just leave the method without any change.

File: pycarol/carolina.py (hunk 2 above)
Comment (JuvenalDuarte) [reply pycarol/carolina.py:60]: same comment as above

File: pycarol/storage.py [hunk 4]
```diff
@@ -22,27 +22,33 @@ def __init__(self, carol):
...
-        else:
-            Carolina.token = None
+        """
+            Check if backend has been reinitialized. If so, also reinit the GCP storage object
+
+        Returns:
+            True if storage object has been reinitialized, false if it is still valid.
+
+        """
+        if self.carolina.init_if_needed():
```
Comment (rafarui) [inline pycarol/storage.py:37]: I think the TODO comment is that you dont need to validate anything here.. If you see `StorageGCPCS` always calls `carolina.get_client()` and this is always checking if it needs to update tokens

File: pycarol/storage.py [hunk 5]
```diff
@@ -147,6 +153,7 @@ def save(
 
 
         """
+        self._init_if_needed()
```
Comment (rafarui) [inline pycarol/storage.py:156]: for all the `_init_if_needed`

If you see `StorageGCPCS` (that is the backend) always calls `carolina.get_client()` and this is always checking if it needs to update tokens. 

This is adding a overhead I think

File: pycarol/carolina.py (hunk 1 above)
Comment (rafarui) [reply pycarol/carolina.py:47]: I did some tests and I was able to get a scenario that this could happen, but it is a very odd one. let's keep like this.

**Review by rafarui — CHANGES_REQUESTED** [review]

**Pull request closed without merging**